src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Field.java

```diff
@@ -41,22 +42,88 @@ public class Field {
   /** Constructor for fields that are named in an InstanceKlass's
       fields array (i.e., named, non-VM fields) */
   Field(InstanceKlass holder, int fieldIndex) {
+    this(holder, fieldIndex, getFieldInfoValues(new CompressedReadStream(holder.getFieldInfoStream().getDataStart()), fieldIndex));
+  }
+
+  private Field(InstanceKlass holder, int fieldIndex, FieldInfoValues values) {
     this.holder = holder;
     this.fieldIndex = fieldIndex;
+    this.values = values;
+    offset = values.offset;
 
-    offset               = holder.getFieldOffset(fieldIndex);
-    genericSignature     = holder.getFieldGenericSignature(fieldIndex);
-
-    name                 = holder.getFieldName(fieldIndex);
+    name = holder.getSymbolFromIndex(values.nameIndex, isInjected());
+    signature = holder.getSymbolFromIndex(values.signatureIndex, isInjected());
+    offset = values.offset;
     id          = new NamedFieldIdentifier(name.asString());
-
-    signature            = holder.getFieldSignature(fieldIndex);
     fieldType   = new FieldType(signature);
+    accessFlags = new AccessFlags(values.accessFlags);
+
+    if (isGeneric()) {
+      genericSignature = holder.getSymbolFromIndex(values.genericSignatureIndex, isInjected());
+    }
+  }
+
```

Comment:
A comment would be nice. Something like "Holder for Field values read in from the field info stream." Maybe a bit more detail than this.